### PR TITLE
implement gettype() return type extension

### PR DESCRIPTION
**Pull request by staabm**
refs https://github.com/phpstan/phpstan/issues/8614#issuecomment-1367891117

### CODE REVIEW

File: src/Type/Php/GettypeFunctionReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
+	{
+		if (count($functionCall->getArgs()) < 1) {
+			return null;
+		}
+
+		$valueType = $scope->getType($functionCall->getArgs()[0]->value);
+
+		$type = TypeTraverser::map($valueType, static function (Type $valueType, callable $traverse): Type {
```
Comment (staabm) [inline src/Type/Php/GettypeFunctionReturnTypeExtension.php]: alternatively we might implement `Type->getType()` or `Type->getTypeString()` instead

File: src/Type/Php/GettypeFunctionReturnTypeExtension.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+			if ($valueType->isNull()->yes()) {
+				return new ConstantStringType('NULL');
+			}
+
+			if ($valueType->isObject()->yes()) {
+				return new ConstantStringType('object');
+			}
+
+			return $valueType;
```
Comment (ondrejmirtes) [inline src/Type/Php/GettypeFunctionReturnTypeExtension.php]: I disagree with this last line. It should probably do `new StringType()` instead.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/GettypeFunctionReturnTypeExtension.php (hunk 2 above)
Comment (staabm) [reply src/Type/Php/GettypeFunctionReturnTypeExtension.php]: good catch, fixed.

File: src/Analyser/TypeSpecifier.php [hunk 3]
```diff
@@ -1419,15 +1419,19 @@ private function processBooleanConditionalTypes(Scope $scope, SpecifiedTypes $le
 	private function findTypeExpressionsFromBinaryOperation(Scope $scope, Node\Expr\BinaryOp $binaryOperation): ?array
 	{
 		$leftType = $scope->getType($binaryOperation->left);
-		$rightType = $scope->getType($binaryOperation->right);
 		if (
 			$leftType instanceof ConstantScalarType
+			&& !$binaryOperation->right instanceof Node\Scalar\String_
```
Comment (ondrejmirtes) [inline src/Analyser/TypeSpecifier.php]: Eh, what's so special about string? Why did you add this change?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/TypeSpecifier.php (hunk 3 above)
Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: its required to make https://github.com/staabm/phpstan-src/blob/1.10.x/tests/PHPStan/Analyser/data/bug-6901.php pass

otherwise - because of switch beeing re-written to comparisons and the new return type extension - `$leftType` and `$rightType`  will resolve to constant strings and the method - as it was before the PR - would not specify the `FuncCall` expression but the `String_`

Comment (ondrejmirtes) [reply src/Analyser/TypeSpecifier.php]: That feels like a wrong placement for such fix.

`gettype` is handled in TypeSpecifier around line 1072. Wouldn't it help to first check for `gettype` condition and only after that to perform the thing on line 1424?

Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: I think we will have the same problem for other functions being used in a switch statement/identical comparison and have a return type extension with constant scalars.

we cannot detect it in line 1072, as the problem is not within the `FuncCall` itself, but the comparison.

see the bug in action in https://phpstan.org/r/d00e29f6-3163-414b-b83d-f591fd1eaf7d
(I have added a separate test-case for it in [e60ff34](https://github.com/phpstan/phpstan-src/pull/2437/commits/e60ff34b745a99fac5d6c7b18a664e3d8473e8ec))

Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: If you prefer we could separate that bugfix in a new PR

Comment (ondrejmirtes) [reply src/Analyser/TypeSpecifier.php]: Yeah https://phpstan.org/r/d00e29f6-3163-414b-b83d-f591fd1eaf7d shouldn't be happening, but I'm not sure if it's the same bug...

Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: ok, turned it into a separate bug: https://github.com/phpstan/phpstan/issues/9404

Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: I had another look at it.

you were right that https://github.com/phpstan/phpstan/issues/9404 does not related to the problem here.

I think the fix is right though and the description in https://github.com/phpstan/phpstan-src/pull/2437/files#r1220239214 is accurate

Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: to simplify the test in question:


```php
<?php

use function PHPStan\Testing\assertType;

/**
 * @param int $y
 */
function foo($y): void
{
	switch (gettype($y)) {
		case "integer":
			assertType('int', $y);
			break;
		default:
			assertType('*NEVER*', $y);
	}
}
```

runs into 
```
------ ------------------------------------ 
  Line   tst.php                             
 ------ ------------------------------------ 
  15     Expected type *NEVER*, actual: int  
 ------ ------------------------------------ 
```

in this PR, when we drop the `... instanceof Node\Scalar\String_` changes

**Comment by ondrejmirtes** [comment]
As I said, I didn't like the changes in TypeSpecifier. I made some changes on 1.10.x, essentially doing this https://github.com/phpstan/phpstan-src/pull/2437#discussion_r1221026739

And now the tests in your PR pass. I'm gonna merge it if the pipeline is going to be green. Thanks.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by staabm** [comment]
Thanks
